Mathlib/Tactic/NormNum/Basic.lean

```diff
@@ -83,6 +85,54 @@ theorem isInt_cast {R} [Ring R] (n m : ℤ) :
       return (.isNegNat rα na q(isInt_cast $a (.negOfNat $na) $pa) : Result q(Int.cast $a : $α))
     | _ => failure
 
+--TODO: clean up
+theorem isRat_ofScientific_of_true [DivisionRing α] [CharZero α] : {m e nm ne d k : ℕ} → {n : ℤ} →
+    IsNat m nm → IsNat e ne → nm = Int.mul k n → Nat.mul k (Nat.succ d) = (10 : ℕ) ^ ne → IsRat ((Rat.ofScientific m true e) : α) n (Nat.succ d)
+  | _, e, _, _, d, _, _, ⟨rfl⟩, ⟨rfl⟩, h₁, h₂ => ⟨
+    invertibleOfNonzero (α := α) <| Nat.cast_ne_zero.2 <| Nat.succ_ne_zero d,
+    by
+      simp [Rat.ofScientific, Rat.normalize_eq_mkRat, h₁, ←h₂]
+      have := invertibleOfNonzero (α := α) <| Nat.cast_ne_zero.2 <| Nat.succ_ne_zero d
+      rw [← mul_right_inj_of_invertible <| (Nat.succ d : α), Rat.mkRat_mul_left, Rat.mkRat_eq]
+      have : Nat.succ d ≠ 0 := by simp
+      norm_cast
+      simp only [mul_invOf_mul_self_cancel']
+      rw [Rat.coe_nat_eq_divInt <| Nat.succ d, Rat.divInt_mul_divInt_cancel, Rat.divInt_one]
+      simp
+      norm_cast
+      have h₃ := h₂.symm ▸ pow_pos (a := 10) (by decide) e
+      exact Nat.pos_iff_ne_zero.1 <| pos_of_mul_pos_left h₃ (by simp)
+    ⟩
+
+--!! Does this belong here?
+instance DivisionRing.instOfScientific [DivisionRing α] [CharZero α] : OfScientific α where
+  ofScientific (m : ℕ) (b : Bool) (d : ℕ) := Rat.ofScientific m b d
```

Comment:
I think this should be out in the main library.  I'm wondering, though, if this will conflict with Rat.ofScientific in the case of α = Rat?  Maybe we need to do something like
- Rat.ofScientific is a `def` not an instance in Std
- in Mathlib, `DivisionRing.instOfScientific` is an instance, and `Rat` gets its OfScientific instance from that
- in other `Std` users, `Rat.ofScientific` is made an instance post-facto.

But don't act on my speculations, let's see what an expert thinks.

Reply:
Such an instance definitely belongs in another file.  I think it would be best if `DivisionRing` extended `OfScientific` (to avoid potential diamonds).

Reply:
There seems to be support for that on zulip as well (linked thread). Though, @digama0, you suggested a blanket implementation for anything that has an instance of `RatCast`. In almost all practical cases, I imagine, this will mean it is a `DivisionRing` anyway—but there could be edge cases I don't know of. Do you have an opinion on adding a field to `DivisionRing` vs. making a blanket instance?

Reply:
I meant something with the instances you need to get `Rat.cast`, not `RatCast` directly but the lawful typeclass that implies it (I think this is `DivisionRing` alone)